## Chapter 2.2 - Exercise 1: Thực hiện những yêu cầu liên quan đến series, và đối chiếu với kết quả cho trước

### Part 1: Thực hiện các phép toán trên series

In [77]:
import numpy as np
import pandas as pd

In [78]:
# Câu 1a: Cho arr_1 là mảng số nguyên chẵn [2, 4, 6, 8, 10], arr_2 là mảng số nguyên lẻ [1, 3, 5, 7, 11]
# Tạo biến kiểu Serries ser1 từ arr_1, ser2 từ arr_2
arr_1 = np.array([2, 4, 6, 8, 10])
arr_2 = np.array([1, 3, 5, 7, 11])
ser1 = pd.Series(arr_1)
ser2 = pd.Series(arr_2)

# In danh sách các phần tử của ser1 và ser2
print(ser1)
print(ser2)

0     2
1     4
2     6
3     8
4    10
dtype: int64
0     1
1     3
2     5
3     7
4    11
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
2     6
3     8
4    10
dtype: int32
0     1
1     3
2     5
3     7
4    11
dtype: int32
</pre>
  
</details>

In [79]:
# Câu 1b: Thực hiện phép toán và thể hiện kết quả của:  ser1 + ser2
print(ser1 + ser2)

0     3
1     7
2    11
3    15
4    21
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     3
1     7
2    11
3    15
4    21
dtype: int32</pre>
  
</details>

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    1
1    1
2    1
3    1
4   -1
dtype: int32</pre>
  
</details>

In [80]:
# Câu 1d: Thực hiện phép toán và thể hiện kết quả của:  ser1 * ser2
print(ser1*ser2)

0      2
1     12
2     30
3     56
4    110
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0      2
1     12
2     30
3     56
4    110
dtype: int32</pre>
  
</details>

In [81]:
# Câu 1e: Thực hiện phép toán và thể hiện kết quả của:  ser1 / ser2
print(ser1/ser2)

0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64</pre>
  
</details>

In [82]:
# Câu 2a: Kiểm tra xem các phần tử của ser1 có > các phần tử của ser2 không?
arrBool = (ser1 > ser2).any()
print(arrBool)

True


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     True
1     True
2     True
3     True
4    False
dtype: bool</pre>
  
</details>

In [83]:
# Câu 2b: Kiểm tra xem các phần tử của ser1 có < các phần tử của ser2 không?
arrBool1 = (ser1<ser2).any()
print(arrBool1)

True


In [84]:
# Câu 2c: Kiểm tra xem các phần tử của ser1 có = các phần tử của ser2 không?
print(ser1 == ser2.iloc[:len(ser1)])

0    False
1    False
2    False
3    False
4    False
dtype: bool


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    False
1    False
2    False
3    False
4     True
dtype: bool</pre>
  
</details>

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    False
1    False
2    False
3    False
4    False
dtype: bool</pre>
  
</details>

In [85]:
# Câu 3a: Thêm 2 phần tử [6, 12] vào ser2
ser2[5] = 6
ser2[6]=12
# In lại danh sách các phần tử của ser2.
print(ser2)

0     1
1     3
2     5
3     7
4    11
5     6
6    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     1
1     3
2     5
3     7
4    11
0     6
1    12
dtype: int64</pre>
  
</details>

In [86]:
# Câu 3b: Tạo series ser3 chỉ chứa các phần tử có trong ser1 mà không có trong ser2.
#         In danh sách các phần tử của ser3
ser1.isin(ser2)
ser3 = ser1[~ser1.isin(ser2)]
np3_1 = np.setdiff1d(ser1,ser2)
ser3_1 = pd.Series(np3_1)
print(ser3)
print(ser3_1)

0     2
1     4
3     8
4    10
dtype: int64
0     2
1     4
2     8
3    10
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
3     8
4    10
dtype: int32
</pre>
  
</details>

In [87]:
# Câu 3c: Tạo series ser4 chỉ chứa các phần tử có trong ser2 mà không có trong ser1.
#         In danh sách các phần tử của ser4
ser4 = ser2[~ser2.isin(ser1)]
print(ser4)

0     1
1     3
2     5
3     7
4    11
6    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>
  
</details>

In [88]:
# Câu 4: Tạo series ser5 chứa các phần tử chỉ có trong ser1 và chỉ có trong ser2
#        In danh sách các phần tử của ser5
ser5 = pd.Series(np.union1d(ser1,ser2))
print(ser5)


0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8     10
9     11
10    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
3     8
4    10
0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>
  
</details>

### Part 2: Truy xuất các phần tử, và thống kê thông tin trên series

In [89]:
# Câu 1a: Tạo series ser6 có 35 phần tử số nguyên ngẫu nhiên có giá trị trong khoảng từ 1 đến 9.
np.random.seed(1)
arr = np.random.randint(1,10,35)
ser6 = pd.Series(arr)
print(ser6)
# Cho biết kích thước (shape) của ser6
print(ser6.ndim)
print(ser6.size)
#

# Xem 5 dòng dữ liệu đầu tiên (head) và 5 dòng dữ liệu cuối cùng (tail) có trong ser6
print(ser6.head(5))
print(ser6.tail(5))

0     6
1     9
2     6
3     1
4     1
5     2
6     8
7     7
8     3
9     5
10    6
11    3
12    5
13    3
14    5
15    8
16    8
17    2
18    8
19    1
20    7
21    8
22    7
23    2
24    1
25    2
26    9
27    9
28    4
29    9
30    8
31    4
32    7
33    6
34    2
dtype: int64
1
35
0    6
1    9
2    6
3    1
4    1
dtype: int64
30    8
31    4
32    7
33    6
34    2
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(35,)
0    6
1    9
2    6
3    1
4    1
dtype: int32
30    8
31    4
32    7
33    6
34    2
dtype: int32
</pre>
  
</details>

In [90]:
# Câu 1b: In danh sách các phần tử của ser6 theo dạng array
print(ser6.values) #c1
ser6.to_numpy() #c2
ser6.to_list()

[6 9 6 1 1 2 8 7 3 5 6 3 5 3 5 8 8 2 8 1 7 8 7 2 1 2 9 9 4 9 8 4 7 6 2]


[6,
 9,
 6,
 1,
 1,
 2,
 8,
 7,
 3,
 5,
 6,
 3,
 5,
 3,
 5,
 8,
 8,
 2,
 8,
 1,
 7,
 8,
 7,
 2,
 1,
 2,
 9,
 9,
 4,
 9,
 8,
 4,
 7,
 6,
 2]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>[6 9 6 1 1 2 8 7 3 5 6 3 5 3 5 8 8 2 8 1 7 8 7 2 1 2 9 9 4 9 8 4 7 6 2]
</pre>
  
</details>

In [91]:
# Câu 1c: Cho biết thông tin thống kê chung (describe()) của ser6
print(ser6.describe())

count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64</pre>
  
</details>

In [92]:
# Câu 1d: Cho biết tổng của các phần tử có trong ser6
print(ser6.sum())

182


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>182</pre>
  
</details>

In [93]:
# Câu 1e: Cho biết phần tử có tần suất xuất hiện nhiều nhất trong ser6
print(ser6.mode())
#dem so lan xuat hien
x = ser6[ser6==8].count()
print(x)

0    8
dtype: int64
6


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    8
dtype: int32</pre>
  
</details>

In [94]:
# Câu 2:  Liệt kê các dòng trong ser6 mà giá trị chia hết cho 2 và cho 3
boolArr = (ser6%2==0) & (ser6%3==0)
print(ser6[boolArr])
#---


0     6
2     6
10    6
33    6
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     6
2     6
10    6
33    6
dtype: int32</pre>
  
</details>

In [95]:
# Câu 3: In các phần tử ở vị trí 0, 5, 10, 15 trong ser6
boolARR1 = [0,5,10,15]
print(ser6[boolARR1])
#hien thi tu index 10 den het
ser6 = ser6[10:]
print(ser6)

0     6
5     2
10    6
15    8
dtype: int64
10    6
11    3
12    5
13    3
14    5
15    8
16    8
17    2
18    8
19    1
20    7
21    8
22    7
23    2
24    1
25    2
26    9
27    9
28    4
29    9
30    8
31    4
32    7
33    6
34    2
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     6
5     2
10    6
15    8
dtype: int32</pre>
  
</details>

In [96]:
# Câu 4: In ra các giá trị unique (array) trong ser6
print(ser6.unique())

[6 3 5 8 2 1 7 9 4]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([6, 9, 1, 2, 8, 7, 3, 5, 4], dtype=int64)</pre>
  
</details>

In [97]:
# Câu 5: Tạo series ser7 với mỗi phần tử có giá trị = lập phương của phần tử trong ser6.
ser7 = ser6.map(lambda x: x**3)
print(ser7)

def cube(x)
 return x**3
 ser7 = ser6.map(cube)
# Xem 5 dòng dữ liệu đầu tiên (head) của ser7
print(ser7.head(5))

10    216
11     27
12    125
13     27
14    125
15    512
16    512
17      8
18    512
19      1
20    343
21    512
22    343
23      8
24      1
25      8
26    729
27    729
28     64
29    729
30    512
31     64
32    343
33    216
34      8
dtype: int64
10    216
11     27
12    125
13     27
14    125
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    216.0
1    729.0
2    216.0
3      1.0
4      1.0
dtype: float64
</pre>
  
</details>

### Part 3: Tạo series từ list, chuỗi và biểu thức điều kiện

In [98]:
# Câu 1: Cho list sau:
lst = ["abc", "defg", "htlmj", "dfg", "ljsac"]

In [99]:
# Câu 1a: Tạo series ser_chuoi từ lst
ser_chuoi = pd.Series(lst)
print(ser_chuoi)

0      abc
1     defg
2    htlmj
3      dfg
4    ljsac
dtype: object


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0      abc
1     defg
2    htlmj
3      dfg
4    ljsac
dtype: object
</pre>
  
</details>

In [104]:
# Câu 1b: Tạo series ser_dodai với mỗi phần tử có giá trị là chiều dài của mỗi phần tử trong ser_chuoi
ser_dodai = ser_chuoi.map(lambda x: len(x))
print(ser_dodai)

0    3
1    4
2    5
3    3
4    5
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    3
1    4
2    5
3    3
4    5
dtype: int64</pre>
  
</details>

In [105]:
# Câu 2: Cho ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9])).
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng trong ser có giá trị là số nguyên tố
ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9]))
def is_prime(n):
  if n < 2:
    return False
  for i in range(2, int(np.sqrt(n)) + 1):
    if n % i == 0:
      return False
  return True
ser_nguyento = ser.map(is_prime)
print(ser[ser_nguyento])

1    2
3    5
5    7
dtype: int64


In [108]:
# Câu 2: Cho ser = pd.Series(np.array(['TB', '>TB', '<TB', '>TB', 'TB', 'TB', '<TB', '>TB'])).
ser_tb = pd.Series(np.array(['TB', '>TB', '<TB', '>TB', 'TB', 'TB', '<TB', '>TB']))
#chuyen doi tp = 1 , <tp = 0 , >tp = 2
def covert(x):
  if x== 'TB':
    return 1
  elif x=='>TB':
    return 2
  else:
    return 0
ser = ser_tb.map(covert)
print(ser)

0    1
1    2
2    0
3    2
4    1
5    1
6    0
7    2
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>1    2
3    5
5    7
dtype: int32
</pre>
  
</details>

In [102]:
# Câu 3: Cho mẫu email như sau:
pattern ='[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,4}'

# Tạo một series ser_ch, với mỗi phần tử trong ser_ch là một chuỗi
# Gợi ý: 'reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn'
arr = np.array(['reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn'])
ser_ch = pd.Series(arr)

# In ra những dòng trong ser_ch thỏa điều kiện chuỗi là email
boolArr = ser_ch.str.contains(pattern)
print(ser_ch[boolArr])

1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object</pre>
  
</details>

In [109]:
# Câu 4: Cho series:
ser_names = pd.Series(['Manufacturer', 'Model', 'CarType', 'Min_Price', 'Price', 'Max_Price',
       'MPG_city', 'MPG_highway', 'AirBags', 'DriveTrain', 'Cylinders',
       'EngineSize', 'Horsepower', 'RPM', 'Rev_per_mile', 'Man_trans_avail',
       'Fuel_tank_capacity', 'Passengers', 'Length', 'Wheelbase', 'Width',
       'Turn_circle', 'Rear_seat_room', 'Luggage_room', 'Weight', 'Origin',
       'Make'])
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng của ser_names thỏa điều kiện trong chuỗi có chữ 'Price'
boolArr = ser_names.str.contains('Price')
print(ser_names[boolArr])
def check(x):
  return 'Price' in x
ser_names = ser_names[ser_names.map(check)]
print(ser_names)


3    Min_Price
4        Price
5    Max_Price
dtype: object
3    Min_Price
4        Price
5    Max_Price
dtype: object


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>3    Min_Price
4        Price
5    Max_Price
dtype: object
</pre>
  
</details>